File: prototype/rails/src/test/actions/listings/create_listing_test.rb
require "commerce_test_case"

module Listings
  class CreateListingTest < CommerceTestCase
    def test_a_new_listing_starts_as_a_draft
      created = CreateListing.new.call(seller: seller, draft: draft)

      assert_equal Domain::Listings::ListingStatus::DRAFT, created.status
    end

    def test_it_stores_the_typed_fields_and_the_price_in_cents
      created = CreateListing.new.call(seller: seller, draft: draft)

      assert_equal "Harbour at Dusk", created.title
      assert_equal "Oil", created.medium
      assert_equal "40 x 60 cm", created.dimensions
      assert_equal 24_900, created.price_cents
      assert_equal 2, created.quantity
    end

    def test_it_belongs_to_the_seller_who_created_it
      artist = seller

      created = CreateListing.new.call(seller: artist, draft: draft)

      assert_equal artist, created.seller
    end

    def test_it_slugs_the_title
      created = CreateListing.new.call(seller: seller, draft: draft)

      assert_equal "harbour-at-dusk", created.slug
    end

    def test_a_title_another_listing_already_slugged_is_numbered
      CreateListing.new.call(seller: seller, draft: draft)

      created = CreateListing.new.call(seller: seller, draft: draft)

      assert_equal "harbour-at-dusk-2", created.slug
    end

    def test_it_attaches_an_uploaded_image
      created = CreateListing.new.call(seller: seller, draft: draft, image: uploaded_image)

      assert_predicate created.image, :attached?
      assert_equal "image/png", created.image.content_type
    end

    def test_a_listing_with_no_upload_carries_no_image
      created = CreateListing.new.call(seller: seller, draft: draft)

      refute_predicate created.image, :attached?
    end

    private

    def draft(**overrides)
      Domain::Listings::ListingDraft.from({
        title: "Harbour at Dusk",
        description: "Oil on canvas.",
        medium: "Oil",
        dimensions: "40 x 60 cm",
        price: "249.00",
        quantity: "2"
      }.merge(overrides))
    end

    def uploaded_image
      Rack::Test::UploadedFile.new(
        StringIO.new("\x89PNG\r\n\x1a\n"), "image/png", true, original_filename: "harbour.png"
      )
    end
  end
end

File: prototype/node/src/app/actions/listings/create-listing.test.ts
import { test } from 'node:test'
import assert from 'node:assert/strict'
import { createListing } from './create-listing.ts'
import { parseListingDraft, type ListingDraft, type ListingDraftFields } from '../../core/listings/listing-draft.ts'
import { createSeller, openCommerceWorld } from '../../test/commerce-world.ts'

test('a new listing starts as a draft', async (t) => {
  const world = await openCommerceWorld()
  t.after(world.close)
  const { context } = world

  const sellerId = await createSeller(context)

  const created = await createListing(context, { sellerId, draft: draft() })

  assert.equal(created.status, 'draft')
})

test('it stores the typed fields with the price in cents', async (t) => {
  const world = await openCommerceWorld()
  t.after(world.close)
  const { context } = world

  const sellerId = await createSeller(context)

  const created = await createListing(context, { sellerId, draft: draft() })

  assert.equal(created.title, 'Harbour at Dusk')
  assert.equal(created.description, 'Oil on canvas.')
  assert.equal(created.medium, 'Oil')
  assert.equal(created.dimensions, '40 x 60 cm')
  assert.equal(created.priceCents, 24_900)
  assert.equal(created.quantity, 2)
})

test('it belongs to the seller who created it', async (t) => {
  const world = await openCommerceWorld()
  t.after(world.close)
  const { context } = world

  const sellerId = await createSeller(context)

  const created = await createListing(context, { sellerId, draft: draft() })

  assert.equal(created.sellerId, sellerId)
})

test('it slugs the title', async (t) => {
  const world = await openCommerceWorld()
  t.after(world.close)
  const { context } = world

  const sellerId = await createSeller(context)

  const created = await createListing(context, { sellerId, draft: draft() })

  assert.equal(created.slug, 'harbour-at-dusk')
})

test('a title another listing already slugged is numbered', async (t) => {
  const world = await openCommerceWorld()
  t.after(world.close)
  const { context } = world

  const sellerId = await createSeller(context)
  await createListing(context, { sellerId, draft: draft() })

  const created = await createListing(context, { sellerId, draft: draft() })

  assert.equal(created.slug, 'harbour-at-dusk-2')
})

test('an imagePath is stored', async (t) => {
  const world = await openCommerceWorld()
  t.after(world.close)
  const { context } = world

  const sellerId = await createSeller(context)

  const created = await createListing(context, {
    sellerId,
    draft: draft(),
    imagePath: '/uploads/harbour.png',
  })

  assert.equal(created.imagePath, '/uploads/harbour.png')
})

test('a listing with no upload carries null', async (t) => {
  const world = await openCommerceWorld()
  t.after(world.close)
  const { context } = world

  const sellerId = await createSeller(context)

  const created = await createListing(context, { sellerId, draft: draft() })

  assert.equal(created.imagePath, null)
})

function draft(overrides: ListingDraftFields = {}): ListingDraft {
  const parsed = parseListingDraft({
    title: 'Harbour at Dusk',
    description: 'Oil on canvas.',
    medium: 'Oil',
    dimensions: '40 x 60 cm',
    price: '249.00',
    quantity: '2',
    ...overrides,
  })
  if (!parsed.ok) throw new Error(`the fixture draft is not valid: ${JSON.stringify(parsed.errors)}`)

  return parsed.value
}
